File: scripts/get-literature.R
# Usage:
# Rscript get-literature.R STRchive-loci.json literature-directory out-citations.json STRchive-loci-literature.json

# More useful error messages
options(error=traceback)

## Set default CRAN mirror
local({r <- getOption("repos")
       r["CRAN"] <- "https://cran.r-project.org"
       options(repos=r)
})

# Install biomaRt if not already installed
tryCatch({
  if (!requireNamespace("biomaRt", quietly = TRUE)) {
    if (!requireNamespace("BiocManager", quietly = TRUE)) {
      install.packages("BiocManager", ask = FALSE, update = FALSE)
    }
    BiocManager::install("biomaRt", ask = FALSE)
  }
}, error = function(e) {
  cat("Error installing biomaRt package.\n", file=stderr())
  quit(status = 1)
})

tryCatch({
suppressPackageStartupMessages({
  library(jsonlite)
  library(dplyr)
  library(biomaRt)
  library(rentrez)
  library(stringr)
  library(purrr)
})
}, error = function(e) {
  cat("Error loading required packages.\n", file=stderr())
  quit(status = 1)
})

### Data Setup
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 4) {
  stop("Need input json, base directory, output json for all citations and output json for citations per locus")
}

cat("Arguments: ", args, "\n", file=stderr())

# Manually downloaded from biomart
synonym_file = 'data/literature/gene_synonyms.tsv' # should make an arg with default?

data <- fromJSON(args[1])

# get unique references used in the json for each entry
data <- data %>%
  mutate(references = apply(data[, c("age_onset", "mechanism_detail", "details",
                                     "prevalence_details", "year", "disease_description")], 1, function(row) {
                                       # Extract content inside square brackets
                                       matches <- unlist(regmatches(row, gregexpr("(?<=\\[)[^\\]]+(?=\\])", row, perl = TRUE)))

                                       # Replace "; " with "," for multiple citations within one
                                       matches <- gsub("; ", ",", matches)

                                       # Collapse matches into a single string, remove duplicates, and trim whitespaces
                                       unique_matches <- unique(trimws(matches))

                                       # Return the cleaned references
                                       paste(unique_matches, collapse = ",")
                                     }))


### Established loci lit  retrieval
# Extract STRchive gene names into a list
gene_list <- as.character(unique(data$gene))
cat("Searching for genes:", gene_list, "\n", file=stderr())

# Filter out NA values, if there are any, from the 'gene' column
filtered_data <- data[!is.na(data$gene), ]

# Extract STRchive gene names into a list
gene_list <- as.character(unique(filtered_data$gene))

### Use biomaRt to get gene synonyms
# This is necessary because gene names have evolved over time and can differ
# by publication
hosts <- c("https://useast.ensembl.org", "https://asia.ensembl.org", "https://www.ensembl.org")
for (host in hosts) {
  tryCatch({
    mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl",
                    host = host, verbose = TRUE)
    break
  }, error = function(e) {
    cat("Failed to create the mart object for host:", host, ". Trying another.\n", file=stderr())
  })
}

# Retrieve the HGNC symbol and Gene Synonym for the genes in the list
# Check if the mart object is created
if (exists("mart")) {
  # Retrieve the HGNC symbol and Gene Synonym for the genes in the list
  gene_info <- getBM(attributes = c("hgnc_symbol", "external_synonym"),
                     filters = "hgnc_symbol",
                     values = gene_list,
                     mart = mart)

} else {
  cat(paste0("Failed to create the mart object. Using previous gene synonyms from file: ", synonym_file), file=stderr())
  gene_info = read.csv(synonym_file, sep = '\t')
  gene_info = merge(gene_info, data.frame(hgnc_symbol = gene_list, external_synonym = NA), all = TRUE)
  gene_info = unique(gene_info)
}

# Sort gene_info by 'hgnc_symbol' to ensure consistency
gene_info <- gene_info[order(gene_info$hgnc_symbol), ]

# Replace missing values in 'external_synonym' with 'hgnc_symbol'
# This is because some gene_names have no synonyms, so it's easier to group this way
gene_info$external_synonym[is.na(gene_info$external_synonym) |
                             gene_info$external_synonym == ""] <- gene_info$hgnc_symbol[is.na(
                               gene_info$external_synonym) | gene_info$external_synonym == ""]

# curated list of synonyms to exclude based on non-specific results
excluded_synonym_list <- c("B37", "MHP", "MED", "DM", "DM1", "FA", "GAC", "SPD",
                           "PRP", "A1", "CCD", "PHP", "VCF", "PEM", "MCD", "EMA")

#remove from gene_info
gene_info <- gene_info %>%
  filter(!grepl(paste(excluded_synonym_list, collapse = '|'), external_synonym))

# adding quotation marks to terms for query
gene_info <- gene_info %>%
  mutate(across(c(external_synonym, hgnc_symbol), ~sprintf('"%s"', .)))

# adding gene synonym missed by biomaRt
if ("FMR1" %in% gene_list) {
  gene_info <- rbind(gene_info, data.frame(hgnc_symbol = "\"FMR1\"", external_synonym = "\"FMR-1\""))
}

#because pubmed hates slashes and these use slashes
if ("NUTM2B-AS1" %in% gene_list) {
  gene_info <- rbind(gene_info, data.frame(hgnc_symbol = "\"NUTM2B-AS1\"", external_synonym = "\"LOC642361/NUTM2B-AS1\""))
}
if ("ATXN10" %in% gene_list) {
  gene_info <- rbind(gene_info, data.frame(hgnc_symbol = "\"ATXN10\"", external_synonym = "\"SCA10\""))
}

# collapse to gene names
gene_names <- unique(gene_info$hgnc_symbol)

# strings for queries
consolidated_strings <- gene_info %>%
  group_by(hgnc_symbol) %>%
  summarize(consolidated_strings = paste(unique(c(hgnc_symbol, external_synonym)), collapse = ' OR ')) %>%
  pull(consolidated_strings)

#substitute to avoid unrelated PMIDs
consolidated_strings <- gsub("BMD", "Becker muscular dystrophy", consolidated_strings)

#where results will be stored
base_directory <- args[2]

# Optional: set NCBI email from environment (recommended by NCBI)
# if (nzchar(Sys.getenv("ENTREZ_EMAIL"))) {
#   set_entrez_email(Sys.getenv("ENTREZ_EMAIL"))
# } else {
#   cat("Note: Set ENTREZ_EMAIL env var to comply with NCBI policies.\n", file=stderr())
# }

# Helper to fetch MEDLINE text in batches and write to file prefix
fetch_medline_and_write <- function(pmids, outfile_prefix, batch_size = 200) {
  if (length(pmids) == 0) return(FALSE)
  medline_texts <- character(0)
  for (i in seq(1, length(pmids), by = batch_size)) {
    chunk <- pmids[i:min(i + batch_size - 1, length(pmids))]
    txt <- tryCatch({
      entrez_fetch(db = "pubmed", id = chunk, rettype = "medline", retmode = "text")
    }, error = function(e) {
      cat("entrez_fetch error:", conditionMessage(e), "\n", file=stderr())
      return(NULL)
    })
    if (!is.null(txt)) medline_texts <- c(medline_texts, txt)
  }
  if (!length(medline_texts)) return(FALSE)
  out_file <- paste0(outfile_prefix, "_batch_01.txt")
  writeLines(medline_texts, con = out_file, useBytes = TRUE)
  return(file.exists(out_file))
}

# function to perform the pubmed query
# Function printout includes gene name and if there are results, confirms
# that a file has been created
perform_pubmed_query <- function(gene_info) {
  file_paths <- list()  # Initialize the list to store all publications
  for (i in seq_along(gene_names)) {
    gene_name <- gene_names[i]
    cat("Processing gene:", gene_name, "\n", file=stderr())
    # Use str_detect to check if gene_name is present in each consolidated string
    idx <- str_detect(consolidated_strings, regex(gene_name, ignore_case = TRUE))

    # Put together the relevant consolidated strings
    or_terms <- paste(consolidated_strings[idx], collapse = ' OR ')
    # Splitting the or_terms into individual terms
    individual_terms <- unlist(strsplit(or_terms, " OR "))

    # Adding [Title/Abstract] to each term
    joined_terms <- paste0(individual_terms, "[Title/Abstract]")
    # Construct the query with organized or_terms
    terms_repeat <- c(
      '"repeat expansion"[Title/Abstract]',
      '"repeat expansions"[Title/Abstract]',
      '"tandem repeat"[Title/Abstract]',
      '"tandem repeats"[Title/Abstract]',
      '"repeat sequence"[Title/Abstract]',
      '"repeat sequences"[Title/Abstract]',
      '"repeat length"[Title/Abstract]',
      '"repeat lengths"[Title/Abstract]',
      '"expansion"[Title]',
      '"expansions"[Title]',
      '"repeats"[Title]'
    )

    terms_gene <- joined_terms  # e.g., "FMR1"[Title/Abstract] OR "FMR-1"[Title/Abstract] ...
    terms_language <- '"English"[Language]'
    terms_disease <- c(
      "disease*[Title/Abstract]",
      "disorder*[Title/Abstract]",
      "syndrome*[Title/Abstract]",
      "patient*[Title/Abstract]",
      "proband*[Title/Abstract]"
    )
    terms_pubtype <- c(
      '"journal article"[Publication Type]',
      '"letter"[Publication Type]',
      '"Case Reports"[Publication Type]'
    )
    terms_exclude <- '"review"[Publication Type]'

    query <- paste(
      "(" , paste(terms_repeat, collapse = " OR "), ")",
      "AND (", paste(terms_gene, collapse = " OR "), ")",
      "AND", terms_language,
      "AND (", paste(terms_disease, collapse = " OR "), ")",
      "AND (", paste(terms_pubtype, collapse = " OR "), ")",
      "NOT", terms_exclude
    )

    # Clean up double spaces
    query <- gsub("\\s{2,}", " ", query)  # Remove double spaces
    gene_name <- gsub('"', '', gene_name)
    out_prefix <- paste0(base_directory, "/", gene_name)

    # print query for debugging
    # cat("PubMed query for gene", gene_name, ":\n", query, "\n")

    # Use rentrez to search and fetch
    tryCatch({
      srch0 <- entrez_search(db = "pubmed", term = query, retmax = 0, use_history = FALSE)
      count <- ifelse(is.null(srch0$count), 0L, as.integer(srch0$count))
      cat("Found", count, "articles for gene:", gene_name, "\n", file=stderr())
      if (count == 0L) {
        cat("Skipping fetch for:", gene_name, "\n", file=stderr())
        next
      }
      retmax <- min(count, 10000L)
      srch <- entrez_search(db = "pubmed", term = query, retmax = retmax)
      ok <- fetch_medline_and_write(srch$ids, out_prefix)
      if (!ok) stop("Failed to write MEDLINE file")
    }, error = function(e) {
      cat("batch pubmed download error: ", conditionMessage(e), "\n", file=stderr())
      quit(status = 1)
    })

    # output file name
    out_file <- paste0(out_prefix, "_batch_01.txt")
    cat(out_file, "\n", file=stderr())

    # Check if the file was created successfully
    cat("Full file path:", out_file, "\n", file=stderr())
    if (file.exists(out_file)) {
      file_paths[[gene_name]] <- out_file
    } else {
      cat("Error: File not found -", out_file, "\n", file=stderr())
    }
  }

  return(file_paths)
}

file_paths <- perform_pubmed_query(gene_info)

#creating list for publications
all_publications <- list()

# Assuming file_paths is a list of file paths
#let's get the files into a dataframe in R
for (gene_name in names(file_paths)) {
  # Append "01.txt" to the file path
  file_path <- paste0(file_paths[[gene_name]])
  cat("Reading file:", file_path, "\n", file=stderr())

  tryCatch({
    # Read the file into a character vector
    current_publications <- readLines(file_path)

    # Append to the overall list
    all_publications[[gene_name]] <- current_publications
  }, error = function(e) {
    cat("Warning: error reading file:", file_path, "\n", file=stderr())
    # Append an empty character vector in case of an error
    all_publications[[gene_name]] <- character(0)
  })
}

# empty list for the publication info
pub_info_list <- list()

extract_citation_info <- function(medline_data_list, gene_name) {
  # Combine the MEDLINE records into a single string
  medline_string <- paste(medline_data_list, collapse = "")

  # Extract PMIDs, dates, and titles
  pmids <- str_extract_all(medline_string, "(?<=PMID- )\\d+")[[1]]
  publication_dates <- str_extract_all(medline_string, "(?<=DP  - )\\d+")[[1]]
  title <- str_extract_all(medline_string, "(?<=TI  - ).+?(?=\\.|\\?)")[[1]]

  # Align vector lengths
  length_diff <- length(pmids) - length(publication_dates)
  if (length_diff > 0) {
    publication_dates <- c(publication_dates, rep(NA, length_diff))
  } else if (length_diff < 0) {
    pmids <- c(pmids, rep(NA, -length_diff))
  }

  length_diff <- length(pmids) - length(title)
  if (length_diff > 0) {
    title <- c(title, rep(NA, length_diff))
  } else if (length_diff < 0) {
    pmids <- c(pmids, rep(NA, -length_diff))
  }

  # Data frame output
  data.frame(
    gene = rep(gene_name, length(pmids)),
    PMID = pmids,
    PublicationDate = publication_dates,
    Title = title,
    stringsAsFactors = FALSE
  )
}

for (gene_name in names(all_publications)) {
  medline_data_list <- all_publications[[gene_name]]
  if (length(medline_data_list) == 0) next  # skip empty files
  pub_info_df <- extract_citation_info(medline_data_list, gene_name)
  pub_info_list[[gene_name]] <- pub_info_df
}

# Combine all the dataframes into a single dataframe
if (length(pub_info_list) == 0) {
  all_pub_info_df <- data.frame(
    gene = character(),
    PMID = character(),
    PublicationDate = character(),
    Title = character(),
    stringsAsFactors = FALSE
  )
} else {
  all_pub_info_df <- do.call(rbind, pub_info_list)
}

# add pubmed search results to literature field for entry
data <- data %>%
  mutate(additional_literature = map_chr(gene, function(g) {
    # Guard if no publications parsed
    if (nrow(all_pub_info_df) == 0) return("")
    matching_pmids <- all_pub_info_df %>%
      filter(gene == g) %>%
      pull(PMID) %>%
      unique() %>%
      paste0("@pmid:", ., collapse = ",")
    if (length(matching_pmids) == 0) "" else matching_pmids
  }))

# remove any redundant pmids from additional_literature that are in references
data <- data %>%
  mutate(additional_literature = mapply(function(lit, refs) {
    # Split the strings by commas to create lists
    lit_list <- unique(unlist(str_split(lit, ",\\s*")))  # Split and remove extra spaces
    refs_list <- unique(unlist(str_split(refs, ",\\s*")))  # Split and remove extra spaces

    # Find common elements between additional_literature and references
    common_elements <- intersect(lit_list, refs_list)

    # Print common elements being removed (if any)
    if (length(common_elements) > 0) {
      cat("Removing the following elements from additional_literature: ", paste(common_elements, collapse = ", "), "\n", file=stderr())
    }

    # Remove common elements from additional_literature
    lit_list <- setdiff(lit_list, common_elements)

    # Join the remaining elements back into a string
    cleaned_lit <- paste(lit_list, collapse = ",")

    # Return the cleaned additional_literature
    return(cleaned_lit)
  }, data$additional_literature, data$references))

# Write out a json with just the locus id and the two literature fields
lit_data <- data[, c("id", "additional_literature", "references")]
write_json(lit_data, args[4])

#### New locus lit retrieval
#new locus query found from reviewing pertinent terms in discovery papers
perform_new_pubmed_query <- function() {
  file_path <- list()  # Initialize the list to store all publications

  terms_repeat <- c(
    '"repeat expansion"[Title/Abstract]',
    '"tandem repeat"[Title/Abstract]'
  )
  terms_discovery <- c(
    '"discovered"[Title/Abstract]',
    '"identified"[Title/Abstract]',
    '"causative"[Title/Abstract]',
    '"underlie"[Title/Abstract]',
    '"basis"[Title/Abstract]'
  )
  terms_language <- '"English"[Language]'
  terms_disease <- c(
    '"disease"[Title/Abstract]',
    '"disorder"[Title/Abstract]',
    '"syndrome"[Title/Abstract]',
    '"condition*"[Title/Abstract]'
  )
  terms_pubtype <- c(
    '"journal article"[Publication Type]',
    '"letter"[Publication Type]',
    '"Case Reports"[Publication Type]'
  )
  terms_exclude <- '"review"[Publication Type]'

  query <- paste(
    "(" , paste(terms_repeat, collapse = " OR "), ")",
    "AND (", paste(terms_discovery, collapse = " OR "), ")",
    "AND", terms_language,
    "AND (", paste(terms_disease, collapse = " OR "), ")",
    "AND (", paste(terms_pubtype, collapse = " OR "), ")",
    "NOT", terms_exclude
  )

  query <- gsub("\\s{2,}", " ", query)
  out_prefix <- paste0(base_directory, "/new_loci")

  tryCatch({
    srch0 <- entrez_search(db = "pubmed", term = query, retmax = 0, use_history = FALSE)
    count <- ifelse(is.null(srch0$count), 0L, as.integer(srch0$count))
    if (count == 0L) {
      cat("Skipping fetch for new loci - no articles found.\n", file=stderr())
      return(NULL)
    }
    retmax <- min(count, 10000L)
    srch <- entrez_search(db = "pubmed", term = query, retmax = retmax)
    ok <- fetch_medline_and_write(srch$ids, out_prefix)
    if (!ok) stop("Failed to write MEDLINE file")
  }, error = function(e) {
    cat("batch pubmed download error: ", conditionMessage(e), "\n", file=stderr())
    quit(status = 1)
  })

  out_file <- paste0(out_prefix, "_batch_01.txt")
  cat("Full file path:", out_file, "\n", file=stderr())
  if (file.exists(out_file)) {
    file_path <- out_file
  } else {
    cat("Error: File not found -", out_file, "\n", file=stderr())
  }

  return(file_path)
}

#
# ### Let's get all the citations to run manubot on
extract_citations <- function(column) {
  column %>%
    str_split(",") %>%                  # Split on commas
    unlist() %>%                        # Flatten list to vector
    str_trim() %>%                      # Trim whitespace
    str_subset("^@\\S+") %>%            # Ensure only terms starting with @ are included
    str_remove("^@") %>%                # Remove the @ symbol from each term
    unique()                            # Remove duplicates
}

# Extract citations from `references` and `additional_literature`
citations_references <- extract_citations(data$references)
citations_additional <- extract_citations(data$additional_literature)

# Combine citations into a single list and remove duplicates
all_citations <- unique(c(citations_references, citations_additional))

myfile = toJSON(all_citations)

write(myfile, args[3])

cat(paste0("Wrote all citations to ", args[3], "\n"), file=stderr())
quit(status = 0)

File: scripts/get-literature.py

#!/usr/bin/env python3
"""
Retrieve PubMed Literature for monthly lit review
Usage:
  python get-literature.py STRchive-loci.json literature-dir out-citations.json out-loci-literature.json
"""

import csv
import json
import os
import re
import subprocess
import sys
import time
import urllib.error
import urllib.parse
import urllib.request

EUTILS = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/"
BIOMART_HOSTS = ["https://useast.ensembl.org", "https://asia.ensembl.org", "https://www.ensembl.org"]
SYNONYM_FILE = "data/literature/gene_synonyms.tsv"  #Backup
SYNONYM_MAX_AGE_DAYS = float(os.environ.get("STRCHIVE_SYNONYM_MAX_AGE_DAYS", 180))

# Locus fields that can have citation markers
REF_FIELDS = ["age_onset", "mechanism_detail", "details", "prevalence_details",
              "year", "disease_description", "detection"]

# Synonyms to exclude
EXCLUDED_SYNONYMS = ["B37", "MHP", "MED", "DM", "DM1", "FA", "GAC", "SPD",
                     "PRP", "A1", "CCD", "PHP", "VCF", "PEM", "MCD", "EMA"]

# Synonyms BioMart misses
EXTRA_SYNONYMS = {"FMR1": "FMR-1", "NUTM2B-AS1": "LOC642361/NUTM2B-AS1", "ATXN10": "SCA10"}

#Repeat terms to inclide 
TERMS_REPEAT = ['"repeat expansion"[Title/Abstract]', '"repeat expansions"[Title/Abstract]',
                '"tandem repeat"[Title/Abstract]', '"tandem repeats"[Title/Abstract]',
                '"repeat sequence"[Title/Abstract]', '"repeat sequences"[Title/Abstract]',
                '"repeat length"[Title/Abstract]', '"repeat lengths"[Title/Abstract]',
                '"expansion"[Title]', '"expansions"[Title]', '"repeats"[Title]']

#Disease terms to inclide 
TERMS_DISEASE = ["disease*[Title/Abstract]", "disorder*[Title/Abstract]", "syndrome*[Title/Abstract]",
                 "patient*[Title/Abstract]", "proband*[Title/Abstract]"]
#types of publications
TERMS_PUBTYPE = ['"journal article"[Publication Type]', '"letter"[Publication Type]',
                 '"Case Reports"[Publication Type]']

TERMS_LANGUAGE = '"English"[Language]'
TERMS_EXCLUDE = '"review"[Publication Type]'

# New locus query
NEW_TERMS_REPEAT = ['"repeat expansion"[Title/Abstract]', '"tandem repeat"[Title/Abstract]']
NEW_TERMS_DISCOVERY = ['"discovered"[Title/Abstract]', '"identified"[Title/Abstract]',
                       '"causative"[Title/Abstract]', '"underlie"[Title/Abstract]',
                       '"basis"[Title/Abstract]']

NEW_TERMS_DISEASE = ['"disease"[Title/Abstract]', '"disorder"[Title/Abstract]',
                     '"syndrome"[Title/Abstract]', '"condition*"[Title/Abstract]']


def log(*args):
    print(*args, file=sys.stderr)


def dedupe(items):
    # unique order
    seen = {}
    for x in items:
        seen.setdefault(x, None)
    return list(seen)


def text_leaves(value):
   # Return all text leaves
    if value is None or isinstance(value, bool):
        return
    if isinstance(value, str):
        yield value
    elif isinstance(value, (int, float)):
        yield str(value)
    elif isinstance(value, list):
        for item in value:
            yield from text_leaves(item)
    elif isinstance(value, dict):
        for item in value.values():
            yield from text_leaves(item)
    else:
        log(f"Ignoring unexpected {type(value).__name__} in reference field")



def http_post(url, params, tries=4):
    for attempt in range(tries):
        try:
            req = urllib.request.Request(
                url, data=urllib.parse.urlencode(params).encode(),
                headers={"User-Agent": "STRchive get-literature.py"})
            return urllib.request.urlopen(req, timeout=180).read().decode("utf-8", "replace")
        except (urllib.error.URLError, OSError) as e:
            if attempt == tries - 1:
                raise
            log(f"  request failed ({e}), retrying...")
            time.sleep(2 ** attempt)


def eutils(endpoint, params):
    params = dict(params)
    if os.environ.get("NCBI_API_KEY"):
        params["api_key"] = os.environ["NCBI_API_KEY"]
    if os.environ.get("NCBI_EMAIL"):
        params["email"] = os.environ["NCBI_EMAIL"]
    params["tool"] = "STRchive"
    text = http_post(EUTILS + endpoint, params)
    time.sleep(0.11 if os.environ.get("NCBI_API_KEY") else 0.34)  # rate limit
    return text



# synonyms

def synonym_file_age_days():
    #Age of synonym file 
    try:
        ts = subprocess.run(["git", "log", "-1", "--format=%ct", "--", SYNONYM_FILE],
                            capture_output=True, text=True, timeout=10).stdout.strip()
        if ts:
            return (time.time() - int(ts)) / 86400
    except Exception:
        pass
    return (time.time() - os.path.getmtime(SYNONYM_FILE)) / 86400


def read_synonym_file(genes):
    # Fallback synonyms from the local TSV
    if not os.path.exists(SYNONYM_FILE):
        sys.exit(f"BioMart is unreachable and there is no fallback file")

    age_days = synonym_file_age_days()
    log("Failed to create the mart object. Using previous gene synonyms from file:", SYNONYM_FILE)
    log(f"  last modified {age_days:.0f} days ago") # To better know when it was last updated
    if age_days > SYNONYM_MAX_AGE_DAYS and not os.environ.get("STRCHIVE_ALLOW_STALE_SYNONYMS"):
        sys.exit(f" {SYNONYM_FILE} is {age_days:.0f} days old (limit "
                 f"{SYNONYM_MAX_AGE_DAYS:.0f})")

    with open(SYNONYM_FILE, encoding="utf-8") as f:
        reader = csv.DictReader(f, delimiter="\t")
        missing = {"hgnc_symbol", "external_synonym"} - set(reader.fieldnames or [])
        if missing:
            sys.exit(f"{SYNONYM_FILE} is missing required column(s): {', '.join(sorted(missing))}")
        rows = [(r["hgnc_symbol"], r["external_synonym"] or "") for r in reader]

    rows.sort(key=lambda r: (r[1], r[0]))
    return dedupe(rows + [(g, "") for g in genes])


def get_synonyms(genes):
    #(hgnc_symbol, external_synonym) pairs 
    xml = ('<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE Query>'
           '<Query virtualSchemaName="default" formatter="TSV" header="0" uniqueRows="0"'
           ' count="" datasetConfigVersion="0.6">'
           '<Dataset name="hsapiens_gene_ensembl" interface="default">'
           f'<Filter name="hgnc_symbol" value="{",".join(genes)}"/>'
           '<Attribute name="hgnc_symbol"/><Attribute name="external_synonym"/>'
           '</Dataset></Query>')

    requested = set(genes)
    for host in BIOMART_HOSTS: #attempt to retrieve rows
        try: 
            text = http_post(host + "/biomart/martservice", {"query": xml}, tries=1)
            if "Query ERROR" in text:
                raise RuntimeError(text.strip().splitlines()[0])
            rows = [line.split("\t") for line in text.splitlines() if line.strip()]
            rows = [(r[0], r[1] if len(r) > 1 else "") for r in rows if r[0]]
            if not rows:
                raise RuntimeError("empty response")
            if not requested.intersection(sym for sym, _ in rows):
                raise RuntimeError("response contained none of the requested gene symbols")
            log(f"Retrieved {len(rows)} synonym rows from {host}") 
            return rows
        except Exception as e:
            log("Failed to create the mart object for host:", host, ". Trying another.")
            log(f"  ({e})")  # to actually log exceptions 

    return read_synonym_file(genes)


def report_synonym_coverage(rows, genes):
    # Log which genes will be searched under only their primary symbol
    have = {sym for sym, syn in rows if syn and syn != sym}
    absent = [g for g in genes if g not in have and g not in EXTRA_SYNONYMS]
    log(f"Synonyms found for {len(genes) - len(absent)}/{len(genes)} genes.")
    if absent:
        log(f"  Primary symbol only (no synonyms): {' '.join(absent)}")


def build_gene_terms(genes):
    # Return (gene_names, consolidated_strings)
    rows = get_synonyms(genes)
    report_synonym_coverage(rows, genes)

    # Genes with no synonym stand in for themselves
    rows = [(sym, syn or sym) for sym, syn in rows]
    rows.sort(key=lambda r: r[0]) 

    excluded = re.compile("|".join(EXCLUDED_SYNONYMS))
    rows = [r for r in rows if not excluded.search(r[1])]

    dropped = [g for g in genes if g not in {sym for sym, _ in rows} and g not in EXTRA_SYNONYMS]
    if dropped:
        log(f"  WARNING: dropped by EXCLUDED_SYNONYMS, will not be searched: {' '.join(dropped)}")

    rows = [(f'"{sym}"', f'"{syn}"') for sym, syn in rows]
    for sym, syn in EXTRA_SYNONYMS.items():
        if sym in genes:
            rows.append((f'"{sym}"', f'"{syn}"'))

    groups = {}
    for sym, syn in rows:
        groups.setdefault(sym, []).append(syn)

    consolidated = [" OR ".join(dedupe([sym] + syns)) for sym, syns in sorted(groups.items())]
    # Substitute to avoid unrelated PMIDs
    consolidated = [c.replace("BMD", "Becker muscular dystrophy") for c in consolidated]

    return dedupe([sym for sym, _ in rows]), consolidated



# PubMed

def build_query(terms_repeat, terms_middle, terms_disease):
    query = " ".join([
        "(", " OR ".join(terms_repeat), ")",
        "AND (", " OR ".join(terms_middle), ")",
        "AND", TERMS_LANGUAGE,
        "AND (", " OR ".join(terms_disease), ")",
        "AND (", " OR ".join(TERMS_PUBTYPE), ")",
        "NOT", TERMS_EXCLUDE,
    ])
    return re.sub(r"\s{2,}", " ", query)


def fetch_medline(query, out_prefix, gene=None):
    # Search PubMed and write medline records to <out_prefix>_batch_01.txt
    try:
        result = json.loads(eutils("esearch.fcgi", {
            "db": "pubmed", "term": query, "retmax": "10000", "retmode": "json"}))["esearchresult"]
    except Exception as e:
        log("batch pubmed download error: ", e, "")
        sys.exit(1)
    pmids = result.get("idlist", [])
    if gene is not None:
        log("Found", result.get("count", 0), "articles for gene:", gene, "") 
        if not pmids:
            log("Skipping fetch for:", gene, "")
            return ""
    else:
        log("Found", result.get("count", 0), "articles for new loci", "") 
        if not pmids:
            log("Skipping fetch for new loci - No articles found.")
            return ""

    chunks, dropped = [], 0
    for i in range(0, len(pmids), 200):
        batch = pmids[i:i + 200]
        try:
            chunks.append(eutils("efetch.fcgi", {"db": "pubmed", "id": ",".join(batch),
                                                 "rettype": "medline", "retmode": "text"}))
        except Exception as e:
            dropped += len(batch)
            log("entrez_fetch error:", e, "")
    if not chunks:
        return ""

    out_file = out_prefix + "_batch_01.txt"
    with open(out_file, "w", encoding="utf-8") as f:
        f.write("\n".join(chunks) + "\n")  # Not sure if this is actually needed but it was the only way I could get it to exactly match R output. 
    if gene is not None:
        log(out_file, "")
    log("Full file path:", out_file, "")
    if not os.path.exists(out_file):
        log("Error: File not found -", out_file, "")
    return "\n".join(chunks)



def main(in_json, lit_dir, out_citations, out_loci_citations):
    log("Arguments: ", in_json, lit_dir, out_citations, out_loci_citations, "")

    with open(in_json, encoding="utf-8") as f:
        loci = json.load(f)

    bracketed = re.compile(r"(?<=\[)[^\]]+(?=\])")
    for locus in loci:
        found = []
        for field in REF_FIELDS:
            for text in text_leaves(locus.get(field)):
                found += bracketed.findall(text)
        locus["references"] = ",".join(dedupe(m.replace("; ", ",").strip() for m in found))

    genes = dedupe(locus["gene"] for locus in loci if locus.get("gene"))
    log("Searching for genes:", " ".join(genes), "")

    gene_names, consolidated = build_gene_terms(genes)

    os.makedirs(lit_dir, exist_ok=True)
    pmids_by_gene = {}
    for gene_name in gene_names:
        gene = gene_name.strip('"')
        log("Processing gene:", gene_name, "") 
        matched = [c for c in consolidated if re.search(gene_name, c, re.IGNORECASE)]
        terms_gene = [t + "[Title/Abstract]" for t in " OR ".join(matched).split(" OR ")]
        query = build_query(TERMS_REPEAT, terms_gene, TERMS_DISEASE)
        medline = fetch_medline(query, os.path.join(lit_dir, gene), gene=gene)
        pmids_by_gene[gene] = dedupe(re.findall(r"(?<=PMID- )\d+", medline))

    # add search results without PMIDs already there
    for locus in loci:
        found = ["@pmid:" + p for p in pmids_by_gene.get(locus.get("gene"), [])]
        cited = set(re.split(r",\s*", locus["references"]))
        redundant = [p for p in found if p in cited]
        if redundant:
            log("Removing the following elements from additional_literature: ", ", ".join(redundant), "")
        locus["additional_literature"] = ",".join(p for p in found if p not in cited)

    with open(out_loci_citations, "w", encoding="utf-8") as f:
        json.dump([{k: locus[k] for k in ("id", "additional_literature", "references")}
                   for locus in loci], f, separators=(",", ":"))

    # papers that maybe have new loci
    fetch_medline(build_query(NEW_TERMS_REPEAT, NEW_TERMS_DISCOVERY, NEW_TERMS_DISEASE),
                  os.path.join(lit_dir, "new_loci"))

    # Every citation
    citations = []
    for field in ("references", "additional_literature"):
        for locus in loci:
            citations += [t.strip()[1:] for t in locus[field].split(",")
                          if re.match(r"@\S+", t.strip())]
    with open(out_citations, "w", encoding="utf-8") as f:
        json.dump(dedupe(citations), f, separators=(",", ":"))
    log(f"Wrote all citations to {out_citations}")


if __name__ == "__main__":
    if len(sys.argv) != 5:
        sys.exit(__doc__)
    main(*sys.argv[1:])
